### PR TITLE
Remove obsolete content after content synchronization

### LINKED ISSUE #1474: The "Used in" still displays that the image is used in a Product after the Product has been deleted
### Preconditions (*)
1. Installed Magento with Adobe Stock Integration
2. Configured  integration in Stores -> Configuration -> System -> Adobe Stock Integration fieldset

### Steps to reproduce (*)
1. Create a Product and add an image from Media Gallery
2. Go to Content - Media Gallery and View the Details of the image that was previously added to the Product
3. Verify that the Details display that the image is **Used In 1 Product**
4. Go to _Catalog - Products_ and **Delete** the previously created Product
5. Perform ``` bin/magento indexer:reindex, bin/magento media-content:sync, bin/magento cache:clean``` 
6. Go to Go to Content - Media Gallery and View the Details of the image that was previously Used

### Expected result (*)
The image is not Used in a product so it should not be displayed in Details

### Actual result (*)
The Details still display that the image is **Used In 1 Product**

### Note
Please also consider
- pages
- blocks
- categories

### PR DESCRIPTION
**Pull request by Nazar65**
### Fixed Issues (if relevant)

1. Fixes magento/adobe-stock-integration#1474: The "Used in" still displays that the image is used in a Product after the Product has been deleted
2. ...

### Manual testing scenarios (*)
1. Create a Product and add an image from Media Gallery
2. Go to Content - Media Gallery and View the Details of the image that was previously added to the Product
3. Verify that the Details display that the image is **Used In 1 Product**
4. Go to _Catalog - Products_ and **Delete** the previously created Product
5. Perform ``` bin/magento indexer:reindex, bin/magento media-content:sync, bin/magento cache:clean``` 
6. Go to Go to Content - Media Gallery and View the Details of the image that was previously Used

The image is not Used in a product so it should not be displayed in Details

### CODE REVIEW

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

File: MediaContentSynchronizationCms/composer.json [hunk 1]
```diff
@@ -8,6 +8,9 @@
         "magento/module-media-gallery-synchronization-api": "*",
         "magento/module-media-content-api": "*"
     },
+    "suggest": {
+         "magento/module-media-content-synchronization": "*"
```
Comment (sivaschenko) [inline MediaContentSynchronizationCms/composer.json]: I think it's worth to introduce the `Magento\MediaContentSynchronizationApi\Model\Entities` class to be able to use the `magento/module-media-gallery-synchronization-api` for configuration. (I.e. the same way we have `\Magento\MediaContentApi\Model\Composite\GetEntityContents` as part of API module)

In this case:
 - `Magento\MediaContentSynchronizationApi\Model\GetEntitiesInterface` should be introduced as an SPI service
 - `Magento\MediaContentSynchronizationApi\Model\Entities` should be introduced for DI configuration (accepting the configuration to the constructor and implementing the `GetEntitiesInterface`)
 - `GetEntitiesInterface` should be used in the `RemoveObsoleteContentAsset`

File: MediaContentSynchronizationCms/etc/di.xml [hunk 2]
```diff
@@ -14,6 +14,22 @@
             </argument>
         </arguments>
     </type>
+    <type name="Magento\MediaContentSynchronization\Model\RemoveObsoleteContentAsset">
+        <arguments>
+            <argument name="entities" xsi:type="array">
+                <item name="cms_block" xsi:type="string">cms_block</item>
+                <item name="cms_page" xsi:type="string">cms_page</item>
+            </argument>
+             <argument name="entityTableNames" xsi:type="array">
```
Comment (sivaschenko) [inline MediaContentSynchronizationCms/etc/di.xml]: I think it's better to use the `Magento\Framework\EntityManager\MetadataPool` to retrieve the table and identifier.
Yes, the `EntityManager` is not the first priority to use, but in this case, I'd say, it's just essential

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the PR @Nazar65 ! Please take a look at my suggestions in the comments

File: MediaContentSynchronization/Model/RemoveObsoleteContentAsset.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    /**
+     * Remove media content if entity already deleted.
+     */
+    public function execute(): void
+    {
+        foreach ($this->entities as $entity) {
+            $columns = ['entity_id', 'entity_type', 'asset_id'];
+            foreach ($this->fetchBatches->execute(self::MEDIA_CONTENT_ASSET_TABLE, $columns) as $batch) {
```
Comment (sivaschenko) [inline MediaContentSynchronization/Model/RemoveObsoleteContentAsset.php]: Is it possible to optimize this operation by building a select that is joining `media_content_asset` to the entity tables and selects the records that are missing a match in the entity table side?

File: MediaContentSynchronizationCms/etc/di.xml (hunk 2 above)
Comment (Nazar65) [reply MediaContentSynchronizationCms/etc/di.xml]: :+1:  ok

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

File: MediaContentSynchronization/Model/RemoveObsoleteContentAsset.php [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        }
+    }
+    
+    /**
+     * Returns records from media_content table in wichs entity_id not exist anymore.
+     *
+     * @param string $entityType
+     * @throws CouldNotDeleteException
+     */
+    private function getRemovedAssets(string $entityType): array
```
Comment (sivaschenko) [inline MediaContentSynchronization/Model/RemoveObsoleteContentAsset.php]: Please extract this method to `Magento\MediaContentSynchronization\Model\ResourceModel\GetOutdatedRelations::execute(...): ContentAssetLinkInterface[]`

File: MediaContentSynchronization/Model/RemoveObsoleteContentAsset.php [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * Remove obsolete asset content links by assets id and entity id.
+     *
+     * @param array $assets
+     */
+    private function deleteObsoleteContentAsset(array $assets): void
+    {
+        try {
+            $this->resourceConnection->getConnection()->delete(
```
Comment (sivaschenko) [inline MediaContentSynchronization/Model/RemoveObsoleteContentAsset.php]: Please use `\Magento\MediaContentApi\Api\DeleteContentAssetLinksInterface`

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for updates @Nazar65 ! Please take a look on my suggestions

Also I think we need a separate PR to magneto2 repo adding observers for MediaContentCatalog and MediaContentCms modules to track deleted entitites

**Comment by Nazar65** [comment]
@magento run all tests

File: MediaContentSynchronization/Model/RemoveObsoleteContentAsset.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    
+    /**
+     * @param DeleteContentAssetLinksInterface $deleteContentAssetLinks
+     * @param GetEntitiesInterface $getEntities
+     * @param GetOutdatedRelations $getOutdatedRelations
+     */
+    public function __construct(
+        DeleteContentAssetLinksInterface $deleteContentAssetLinks,
+        GetEntitiesInterface $getEntities,
+        GetOutdatedRElations $getOutdateRelations
```
Comment (sivaschenko) [inline MediaContentSynchronization/Model/RemoveObsoleteContentAsset.php]:
```suggestion
        GetOutdatedRelations $getOutdateRelations
```

**Review by sivaschenko — COMMENTED** [review]
Looks good!

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by lenaorobei** [comment]
Will be delivered to 2.0.0-develop

**Pull request closed without merging**